### ISSUE TITLE
config: Define tastes in [...:taste] sections

### ISSUE THREAD

**Issue by userzimmermann** [coala/coala#4511]
Continuation of https://github.com/coala/coala/pull/4397 , according to https://github.com/coala/coala/pull/4397#pullrequestreview-51603474

Relates to https://gitlab.com/coala/GSoC-2017/issues/378

```
[all]
files = **.py
aspects = coalaCorrect, shortlog.colonexistence
language = Python 3.6

[all:taste]
shortlog_colon = false
```

@sils @sims1253 @jayvdb @Udayan12167 @coala/aspects-developers

**Comment by sils** [comment]
@userzimmermann do we really want tastes to be in a different section? It feels so disconnected. They belong together with the actual settings IMO

**Comment by Makman2** [comment]
Agreeing with @sils. Why do you want to separate the tastes anyway? :O

**Comment by sils** [comment]
see also https://github.com/coala/cEPs/blob/master/cEP-0005.md#mockup

Sincerely,

Lasse Schuirmann

lasse@schuirmann.net
http://coala.io/ - http://viperdev.io/ - http://gitmate.com/

**Comment by jayvdb** [comment]
I also queried @userzimmermann about this [in gsoc-adhika](https://gitter.im/coala/gsoc-adhika?at=59728c95bc4647297412ac58).
I wasnt sure how this would work in complex configurations , and I think it makes subsections more complex, when there are sibling subsections with non-taste config.

**Comment by userzimmermann** [comment]
@sils @Makman2 Just to make it look clearer! Instead of merging more or less mandatory fixed options with much more flexible kind-of-sublevel options. What is taste, what not? Is that always intuitively visible?

Tastes are also merged with old-style settings! Even more confusing... And taste names should also be verified. If they really exist... So what is old-style setting, what is taste? Do we have any way to check old-style settings? If they really exist...

And why not name some future taste like some let's-say-top-level option? When the name just fits... A separate section would avoid naming conflicts

We can even create separate sections for single aspects. Avoiding the need for too many taste with explicit aspect prefixes for name conflict avoidance 

```ini
[python]
# main options

[python:Formatting]
# Formatting tastes
```

@jayvdb

**Comment by Makman2** [comment]
> So what is old-style setting, what is taste? 

Imo that doesn't matter for the user.

I actually thought it would roughly look like this:
```
[python]
aspects = whitespaces
whitespaces.remove_trailing = true
whitespaces.spaces_kind = tabs

# for multiple assignments to the same aspect (maybe we develop another syntax,
# but I'm more in for a syntax feature than using multiple sections):
whitespaces
    .remove_trailing = true
    .spaces_kind = tabs
```

**Comment by userzimmermann** [comment]
@Makman2 Custom sub-level parsing in INI syntax? Many started like this and ended up in chaos ;) Everybody with their own sub-formats. INI is not suitable for this. There will also be no editor support. So finally people switched to other formats like YAML, or recently TOML, which will also be used for future Python package config. Then let's better also switch...

**Comment by Makman2** [comment]
> Not only the author of a .coafile should be able to understand it ;)

That's right, however the user doesn't need to distinct whether it's an old-style setting or a taste, as long as it works and looks the same (what it probably will).

> So finally people switched to other formats like YAML, or recently TOML, which will also be used for future Python package config. Then let's better also switch...

That's a considerable option :+1:

**Comment by userzimmermann** [comment]
@Makman2 
> That's right, however the user doesn't need to distinct whether it's an old-style setting or a taste, as long as it works and looks the same (what it probably will).

Probably... ;) Anyway... The user should want to get informed about mistyped settings and aspects. It's already kind of inconvenient that coala doesn't check for mistyped setting names. That can make it hard to trace down such config issues when things are not working as expected. And when me mix them with taste it only gets worse. Of course can easily verify taste names. But for the options that don't exactly map to taste names we must assume they are old-style settings. And we can't just warn about all of them, since we can't distinguish between correct and mistyped setting names...

So let's really consider switching to TOML! Nice that you like the idea :) It's a kind of superset of INI format. Current `.coafile`s should still work. The structure can then be similar to this proposal here, but with indentation. Or we create more complex structure in the main section itself. TOML is flexible...

@sils @sims1253 @fneu @jayvdb @Udayan12167 @coala/aspects-developers Your opinions about TOML?

**Comment by fneu** [comment]
I like the idea of using something that has some widespread acceptance, documentation, robust parsers etc.

**Comment by jayvdb** [comment]
I raised TOML with @sils and @Udayan12167 when they were doing the latest rewrite, and the idea was shot down as the new `.coafile` syntax is little less complicated than toml, especially for section headers and some value assignments in toml are quite non-coder-unfriendly.

**Comment by userzimmermann** [comment]
@Makman2 @jayvdb When considering custom INI syntax extensions... My proposal would be:

```ini
[python]
aspects =
    WhiteSpace:
        allow_trailing_space = no
        indent_space = tabs
        ...
    ...
```

**Comment by Makman2** [comment]
@jayvdb do you have an example? I just read an example in the TOML-Readme, and it looks quite simple.

**Comment by jayvdb** [comment]
The biggest regression in easy of use for non-programmers is our file format doesnt require quotes for strings.

The TOML binary values are also quite constricted , compared to the many ways our boolean settings may be set.

The array syntax is another one, forcing use of `[..,..,..]` instead of our current `..,..,..`.

And subsections require tables, which are more complex than our current file format.

**Comment by sims1253** [comment]
@userzimmermann again, please stop pinging me on random PRs/issues

**Comment by Makman2** [comment]
> The biggest regression in easy of use for non-programmers is our file format doesnt require quotes for strings.

Ah yeah that's really bad...

So it's really more for programmers^^

**Comment by Makman2** [comment]
I mean coala is for programmers, but just writing out values like we can now is definitely easier than TOML^^

**Comment by Makman2** [comment]
What about YAML?

**Comment by fneu** [comment]
> strings, bools, lists

While our current format is very simple, it is also very ambiguous. I *believe*, and I may well be wrong, that we also confuse people because they are unsure how a setting is interpreted. I.e., "is `..,..,..` a correct list or just a string containing commas"? (It may indeed be both).

By adhering to a well known standard like TOML, YAML or even JSON we get rid of that ambiguity. 

Also I understand that our current system is not complex enough for what we need. I figure custom INI extensions might be even more confusing.

**Comment by userzimmermann** [comment]
@Makman2 YAML looks so unfriendly! Our cute coala bears don't feel comfortable in the YANG jungle ;)  I also withdraw my YAML proposal :D 

@fneu Please no JSON! :P 

I think, especially for consistency and backwards compatibility, custom INI and TOML are the only reasonable options

**Comment by fneu** [comment]
@userzimmermann I don't want to propose JSON ^^

Just from the looks of it, TOML is just a human readable form of it

**Comment by Makman2** [comment]
I'm also not for JSON, it's oververbose for our purposes (it's imo generally a bit oververbose).

But why does YAML look so unfriendly?
```
section_a:
    ASPECT_SPACING:
        trailing_whitespaces: no
        some_other_value: yes
# or similar
```
I just skimmed over the wiki page for YAML, there are some unintuitive syntax features, but usually it looks not so bad for what we normally need^^

**Comment by Makman2** [comment]
> I think, especially for **consistency and backwards compatibility**, custom INI and TOML are the only reasonable options

That's what sadly a custom INI is not solving ;)

**Comment by userzimmermann** [comment]
@Makman2 

> That's what sadly a custom INI is not solving ;)

Custom INI keeps future `.coafile`s consistent and backwards compatible with current `.coafile`s ;)

Why not use YAML as custom INI extension? :D By generally keeping our parser but running each option value through the YAML parser! On the value's top level we can then also keep our simple `li, s, t` syntax ...

```ini
[python]
files = setup.py, coalib/**.py
aspects =
    WhiteSpace:
        allow_trailing_space: no
        indent_space: tabs
        ...
    ...
```

**Comment by Makman2** [comment]
> Custom INI keeps future .coafiles consistent and backwards compatible with current .coafiles ;)

Ah kk, that's how it's meant ;)

> Why not use YAML as custom INI extension?

Is an idea, however as a user I would ask immediately following question: Where's the difference between `=` and `:`? And then I would get confused. We shouldn't mix "too different" syntaxes.

**Comment by jayvdb** [comment]
A critical problem wrt this discussion is that taste settings must be solved in three weeks for a gsoc project, and I suspect we are continuing to support `.coafile` custom ini for at least the next version.  So logic says other formats are not relevant to this issue.  I've raised https://github.com/coala/coala/issues/4609 to deal with other formats in the future.

I am definitely against yaml inside ini for tastes.   yaml is a extremely large spec, and the two formats would bump up against each other in unusual ways, so we would have a yaml subset only, and then we need to document that.

tastes need to use the existing custom ini format, at least for the first release.

**Comment by sils** [comment]
I agree we shouldn't get distracted by discussing file formats. IMO we keep the existing concepts: meaning tastes are in the same section as the aspects, that's specced in the cEP and it's fully backwards compatible, doesn't need a new design and makes us reach our goals.

**Comment by jayvdb** [comment]
ping @userzimmermann , did you have any other adjustments to the taste configuration beyond what is in the cEP, that we should consider?
Or can this be closed?  (#4609 covers adding another file format, which could provide a more organised layout of config)

**Comment by jayvdb** [comment]
ping @userzimmermann ...?
Re-open if we have missed something.
